File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.java
import java.io.*;
import java.util.*;

public class Main {
	
	static int n, ans = 1000000000;
	static int[][] taste;
	
	public static void dfs(int dep, int s, int b) {
		if (dep == n) {
			if (s != 0 && ans > Math.abs(s - b)) {
				ans = Math.abs(s - b);
			}
			return;
		}
		
		int sv = taste[dep][0], bv = taste[dep][1];
		dfs(dep + 1, s == 0 ? sv : s * sv, b + bv);
		dfs(dep + 1, s, b);
	}
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		n = Integer.parseInt(br.readLine());
		taste = new int[n][2];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			taste[i][0] = Integer.parseInt(st.nextToken());
			taste[i][1] = Integer.parseInt(st.nextToken());			
		}
		dfs(0, 0, 0);
		System.out.println(ans);
	}
	
}

File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.py
import sys
input = sys.stdin.readline

def dfs(dep, s, b):
    global ans
    if dep == n:
        if s != 0 and ans > abs(s - b):
            ans = abs(s - b)
        return
    dfs(dep + 1, s * sb[dep][0] if s else sb[dep][0], b + sb[dep][1])
    dfs(dep + 1, s, b)

n = int(input())
sb = [[*map(int, input().split())] for _ in range(n)]
ans = 1000000000
dfs(0, 0, 0)
print(ans)